napari/_qt/_qplugins/_qnpe2.py

```diff
@@ -0,0 +1,181 @@
+"""
+from __future__ import annotations
+
+import inspect
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+from app_model import Action
+from app_model.types import ToggleRule
+from magicgui.type_map._magicgui import MagicFactory
+from magicgui.widgets import FunctionGui, Widget
+from qtpy.QtWidgets import QWidget
+
+from napari._app_model.constants import MenuGroup, MenuId
+from napari.plugins import menu_item_template
+from napari.plugins._npe2 import (
+    _get_contrib_parent_menu,
+    get_widget_contribution,
+)
+from napari.utils.translations import trans
+from napari.viewer import Viewer
+
+if TYPE_CHECKING:
+    from npe2.manifest import PluginManifest
+    from npe2.types import WidgetCreator
+
+    from napari._qt.qt_main_window import Window
+
+
+def _get_widget_viewer_param(
+    widget_callable: WidgetCreator, widget_name: str
+) -> str:
+    """Get widget parameter name (if any) and check type."""
+    if inspect.isclass(widget_callable) and issubclass(
+        widget_callable,
+        (QWidget, Widget),
+    ):
+        widget_param = ""
+        # Inspection can fail when adding to bundle as it thinks widget is
+        # a builtin
+        try:
+            sig = inspect.signature(widget_callable.__init__)
+        except ValueError:
+            pass
+        else:
+            for param in sig.parameters.values():
+                if param.name == 'napari_viewer':
+                    widget_param = param.name
+                    break
+                if param.annotation in (
+                    'napari.viewer.Viewer',
+                    Viewer,
+                ):
+                    widget_param = param.name
+                    break
+
+    elif isinstance(widget_callable, MagicFactory) or inspect.isfunction(
+        widget_callable
+    ):
+        widget_param = ""
+    else:
+        raise TypeError(
+            trans._(
+                "'{widget}' must be `QtWidgets.QWidget` or `magicgui.widgets.Widget` subclass, `MagicFactory` instance or function. Please raise an issue in napari GitHub with the plugin and widget you were trying to use.",
+                deferred=True,
+                widget=widget_name,
+            )
+        )
+    return widget_param
+
+
+def _get_widget_callback(
+    napari_viewer: Viewer,
+    mf: PluginManifest,
+    widget_name: str,
+    name: str,
+) -> Optional[Tuple[Union[FunctionGui, QWidget, Widget], str]]:
+    """Toggle if widget already built otherwise return widget.
+
+    Returned widget will be added to main window by a processor.
+    Note for magicgui type widget contributions, `Viewer` injection is done by
+    `magicgui.register_type`.
+    """
+    window = napari_viewer.window
+    if name in window._dock_widgets:
+        dock_widget = window._dock_widgets[name]
+        if dock_widget.isVisible():
+            dock_widget.hide()
+        else:
+            dock_widget.show()
+        return None
```

Comment:
Just a nitpick suggestion
```suggestion
    if dock_widget := window._dock_widgets.get(name):
        dock_widget.setVisible(not dock_widget.isVisible())
        return None
```

Reply:
nice!

Reply:
Is `setVisible` do the same as `hide`/`show` ? I didnt know

Reply:
I see it is, good to know. I wasn't sure, so kept the `hide`/`show`